src/controller/CommissioningWindowOpener.h

```diff
@@ -0,0 +1,172 @@
+     *   device will exit commissioning mode after a successful commissioning,
+     *   or after the given `timeout` time.
+     *
+     * @param[in] deviceId      The device Id.
+     * @param[in] timeout       The commissioning mode should terminate after this much time.
+     * @param[in] iteration     The PAKE iteration count associated with the PAKE Passcode ID and ephemeral
+     *                          PAKE passcode verifier to be used for this commissioning.
+     * @param[in] discriminator The long discriminator for the DNS-SD advertisement.
+     * @param[in] setupPIN      The setup PIN to use, or NullOptional to use a randomly-generated one.
+     * @param[in] callback      The function to be called on success or failure of opening of commissioning window.
+     * @param[out] payload      The setup payload, not including the VID/PID bits, that is generated based on the
+     *                          passed-in information.
+     *
+     * @param[in] readVIDPIDAttributes Should the API internally read VID and PID from the device while opening the
+     *                                 commissioning window.  If this argument is `true`, the API will read VID and
+     *                                 PID from the device and include them in the setup payload passed to the
+     *                                 callback.
+     */
+    CHIP_ERROR OpenCommissioningWindow(NodeId deviceId, System::Clock::Seconds16 timeout, uint32_t iteration,
+                                       uint16_t discriminator, Optional<uint32_t> setupPIN,
+                                       Callback::Callback<OnOpenCommissioningWindow> * callback, SetupPayload & payload,
+                                       bool readVIDPIDAttributes = false);
+
+private:
+    enum class Step : uint8_t
+    {
+        // Ready to start opening a commissioning window.
+        kAcceptCommissioningStart,
+        // Need to read VID.
+        kReadVID,
+        // Need to read PID.
+        kReadPID,
+        // Need to open commissioning window.
+        kOpenCommissioningWindow,
+    };
+
+    CHIP_ERROR OpenCommissioningWindowInternal(OperationalDeviceProxy * device);
+    static void OnPIDReadResponse(void * context, uint16_t value);
+    static void OnVIDReadResponse(void * context, VendorId value);
+    static void OnVIDPIDReadFailureResponse(void * context, CHIP_ERROR error);
+    static void OnOpenCommissioningWindowSuccess(void * context, const app::DataModel::NullObjectType &);
+    static void OnOpenCommissioningWindowFailure(void * context, CHIP_ERROR error);
+    static void OnDeviceConnectedCallback(void * context, OperationalDeviceProxy * device);
+    static void OnDeviceConnectionFailureCallback(void * context, PeerId peerId, CHIP_ERROR error);
+
+    // TODO: Salt should be provided as an input or it should be randomly generated when
+    // the PIN is randomly generated.
+    static ByteSpan GetSPAKE2Salt();
```

Comment:
If multiple windows get opened, this will clash across openers. This cannot be `static`.

Reply:
Again, this is pre-existing, as is the TODO comment.  I just moved this code.